pydens/model_torch.py

```diff
@@ -0,0 +1,221 @@
+
+    def forward(self, *xs):
+        xs = super().forward(*xs)
+        u = self.conv_block(xs)
+        return self.anzatc()(u, xs)
+
+def D(y, x):
+    """ Differentiation token.
+    """
+    res = grad(y.sum(), x, retain_graph=True, create_graph=True)[0]
+    return res
+
+def V(name, *args, **kwargs):
+    """ Token for a trainable variable.
+    """
+    # If the variable does not exist yet - create it and register in the model.
+    # Alternatively, fetch the variable from the model - if already created.
+    model = current_model.get()
+    if not hasattr(model, name):
+        setattr(model, name, nn.Parameter(*args, **kwargs))
+    return getattr(model, name)
+
+
+class Solver():
+    """ Solver-class for handling differential equations with neural networks.
+    """
+    def __init__(self, equation, model=ConvBlockModel, constraints=None, **kwargs):
+        self.equation = equation
+        if constraints is None:
+            self.constraints = ()
+        elif isinstance(constraints, (tuple, list)):
+            self.constraints = constraints
+        else:
+            self.constraints = (constraints, )
+        self.losses = []
+        self.optimizer = None
+
+        # Initialize neural network for solving the equation.
+        self.model = model(**kwargs)
+
+        # Bind created model to a global context variable.
+        current_model.set(self.model)
+        self.ctx = copy_context()
+
+        # Perform fake run of the model to create all the variables (coming from V-token).
+        xs = [torch.rand((1, 1)) for _ in range(self.model.total)]
+        for x in xs:
+            x.requires_grad_()
+        u_hat = self.model(*xs)
+        _ = self.ctx.run(self.equation, u_hat, *xs)
+
+
+    def fit(self, niters, batch_size, sampler=None, losses='equation', optimizer='Adam', criterion=nn.MSELoss(),
+            lr=0.001, **kwargs):
+        """ Fit the model inside the solver-instance. """
+        # Initialize the optimizer if supplied.
+        if optimizer is not None:
+            self.optimizer = getattr(torch.optim, optimizer)([p for p in self.model.parameters()
+                                                              if p.requires_grad],
+                                                              lr=lr, **kwargs)
+
+        # Perform `niters`-iterations of optimizer steps.
+        self.model.train()
+        for _ in tqdm(range(niters)):
+            self.optimizer.zero_grad()
+
+            # Sample batch of points and compute solution-approximation on the batch.
+            if sampler is None:
+                xs = [torch.rand((batch_size, 1)) for _ in range(self.model.total)]
+            else:
+                xs_concat = sampler.sample(batch_size).astype(np.float32)
+                xs = [torch.from_numpy(xs_concat[:, i:i+1]) for i in range(xs_concat.shape[1])]
+            for x in xs:
+                x.requires_grad_()
+            u_hat = self.ctx.run(self.model, *xs)
```

Comment:
Making one `(B, N)-shaped` tensor would make your life a lot easier: no weird starred expressions all over the place

Reply:
Fix it in most places - when applying `model`.

Still needed to run `equation` and `constraint` so left it there.